Fix/38233 webauth preload fido

nikosdion opened the pull request:
Pull Request for Issue #38233 and #38654 .

### Summary of Changes

* Use a copy of `fido.jwt` supplied with Joomla in the `plugins/system/webauthn` folder.
* Script to update this file in the `build` directory.
* Update of `build/build.php` to automatically run this script when building a new version of Joomla.
* Integrate the lazy loading of the MDS information from #38661 

### Testing Instructions

You need to have a host where you see a performance loss when the webauth system plugin is activated e.g. blackhole the domain `mds.fidoalliance.org`.

### Actual result BEFORE applying this Pull Request

Slow page load (+5 seconds than usual).

### Expected result AFTER applying this Pull Request

Fast page load.

### Documentation Changes Required

None.

Commit:
Use cached FIDO metadata

Places a copy of fido.jwt in the plugin's directory

Commit:
Use cached FIDO metadata

Use the cached FIDO metadata in the plugin

Commit:
Use cached FIDO metadata

Remove unused local file fallback

Commit:
Use cached FIDO metadata

Simplify

Commit:
Use cached FIDO metadata

Update Joomla build script

Commit:
Use cached FIDO metadata

Let's not break the build if the FIDO server is down

richard67 commented:
@nikosdion Does this PR fix #38662 , too?

nikosdion commented:
Yes, it does! Since we no longer go through the network there is no performance penalty when the FIDO server is down.

Fedik commented:
One performance issue remains, but now it more "local one" so probably not much critical.

Use of `json_decode(json_encode($entry->metadataStatement), true)`.
Because `$entry->metadataStatement` contain base64 encoded icon, it slows down `json_decode`, and so whole thing.

nikosdion commented:
@Fedik But it no longer happens on every request. It only happens when you are registering an authenticator or viewing a list of authenticators. In these instances a total page load performance impact between 7 and 30 milliseconds is perfectly acceptable.

wilsonge commented:
Can I suggest we distribute the ability to update the file over CLI too within the CMS as a command. So people can choose to update the local file too? Understand it will always get overridden by update - but feels relatively acceptable.

Fedik commented:
> But it no longer happens on every request.

yes right, that what I meant with: "it now local" :)

nikosdion commented:
@wilsonge If you're on CLI does running something like `php /path/to/cli/joomla.php core:webauthn:update` make more sense than `wget "https://mds.fidoalliance.org" -O /path/to/plugins/system/webauthn/fido.jwt`? We really just retrieve and store a file, we don't process it any further.

nikosdion commented:
@Fedik Removing that double encode-decode step will cause fatal errors on PHP 8 with some authenticators. It would also only remove 1-5 milliseconds. Too big a risk for too little gains?

Fedik commented:
> It would also only remove 1-5 milliseconds

Sometimes it hard to stop to make it faster, when see that it can be faster :smile: 

> cause fatal errors on PHP 8 with some authenticators

That interesting.
I would probably then just unset icon (and set it back if it used somwhere in code)
```php
$array = $entry->metadataStatement;
unset($array['icon']);
$array = json_decode(json_encode($array), true);
```

But it is not related to current PR.

Commit:
Use cached FIDO metadata

Do not include the fido.jwt in the repo

laoneo commented on build/build.php:
```diff
@@ -305,6 +305,16 @@ function clean_composer(string $dir)
+// Try to update the fido.jwt file
+require_once __DIR__ . '/update_fido_cache.php';
+
+if (!file_exists(rtrim($fullPath, '\\/') . '/plugins/system/webauthn/fido.jwt'))
```
Can this be moved after composer install and then 309 removed? Like that we do not fetch the script twice.

nikosdion replied:
Will do. This code was added before I added the composer script.

Commit:
Use cached FIDO metadata

Make sure not to re-fetch the fido.jwt file if it already exists.

Commit:
Use cached FIDO metadata

Move the check for the fido.jwt file right after composer install, do not try to fetch the file twice

wilsonge commented:
> @wilsonge If you're on CLI does running something like php /path/to/cli/joomla.php core:webauthn:update make more sense than wget "https://mds.fidoalliance.org" -O /path/to/plugins/system/webauthn/fido.jwt? We really just retrieve and store a file, we don't process it any further.

On a VM it's fine - in containers it's more complicated because alpine doesn't ship with wget for instance - so having it "php native" can make things easier in some environments.

brianteeman commented:
do we need to add administrator/cache/fido.jwt to the list of files removed on upgrade?

richard67 commented:
> do we need to add administrator/cache/fido.jwt to the list of files removed on upgrade?

@brianteeman Seems so. But as said many times before, we don't bother the authors of PRs with that and leave that to the release leads (or their helpers like I am one).

richard67 commented:
P.S.: Question is if it's safe to remove it on update or if it still might be used during that process.

nikosdion commented:
It is safe to remove, yes.

laoneo commented:
@nikosdion are the testing instructions still valid? Guess now is only a composer install needed or the drone build should work out of the box? I would like to get here some tests as it looks like all the issues are ironed out.

laoneo commented on administrator/language/en-GB/plg_system_webauthn.ini:
```diff
@@ -24,7 +24,7 @@ PLG_SYSTEM_WEBAUTHN_ERR_NO_BROWSER_SUPPORT="Sorry, your browser does not support
 PLG_SYSTEM_WEBAUTHN_ERR_USER_REMOVED="The user for this authenticator seems to no longer exist on this site."
 PLG_SYSTEM_WEBAUTHN_ERR_XHR_INITCREATE="Cannot get the authenticator registration information from your site."
-PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC="Only allow authenticators with verifiable cryptographic signatures to be used for WebAuthn logins. Strongly recommended for high security environments. Requires your site to be able to access <code>https://mds.fidoalliance.org/</code> directly, a writeable cache directory, the system temporary directory being writeable by PHP, and the OpenSSL ex ...
+PLG_SYSTEM_WEBAUTHN_FIELD_ATTESTATION_SUPPORT_DESC="Only allow authenticators with verifiable cryptographic signatures to be used for WebAuthn logins. Strongly recommended for high security environments. Requires your site to be able to access <code>https://mds.fidoalliance.org/</code> directly, a writeable cache directory, the system temporary directory being writeable by PHP, and the OpenSSL ex ...
```
I guess this needs an update as well as access to fido is not a requirement anymore.

nikosdion replied:
Addressed in https://github.com/joomla/joomla-cms/pull/38664/commits/d51adca6d1691bc0c684fac387b89a657a7b33a1

nikosdion commented:
@laoneo If the FIDO service becomes available then, yes, the testing instructions will work using the Drone build.

Since FIDO servers are down you need to manually copy fido.jwt into plugins/system/webauthn before testing. That's what I was telling in the chat yesterday. The FIDO server seems down so we have to do that step manually when building 4.2.2.

Commit:
Use cached FIDO metadata

Update lang string

Commit:
Use cached FIDO metadata

Set the attestation support to disabled by default

roland-d commented:
@nikosdion Not sure what is going on but when I run the `php build/build.php` the `update_fido_cache.php` is not run. Do I run `composer install` manually, it is triggered.

Does the `php build/build.php` work for you? 

I have a workaround to just copy the fido.jwt from my local system in the build.php but it seems odd that the composer install from the build script does not trigger the `post-install-cmd`

nikosdion commented:
This is what I was talking about with @HLeithner and @laoneo. Here's the problem we have.

For the update_fido_cache.php script to work we need the FIDO server to be working. Otherwise it just hangs for 5 seconds (the timeout we have) and tells you it cannot fetch the file.

The FIDO server is _not_ working at the moment.

They way I see it is that the only way to actually build Joomla 4.2.2 is to commit fido.jwt in the repo (for now!), build the release and remove the file from the repo when the FIDO server starts working again.

roland-d commented:
> The FIDO server is not working at the moment.

That must have been a fluke on my end then because when I called the URL I was given a jwt file :)

> They way I see it is that the only way to actually build Joomla 4.2.2 is to commit fido.jwt in the repo (for now!), build the release and remove the file from the repo when the FIDO server starts working again.

I did not commit the file to the repo, just added a copy statement in the build script to copy it after composer has run. This seems fine to me and less work than committing the file unless you see an issue with that.

nikosdion commented:
@roland-d No problem from me! Your solution is not just acceptable, it's tested. It's actually how I was running my local test as you can see in my commit history on this PR :)

richard67 commented:
> I did not commit the file to the repo, just added a copy statement in the build script to copy it after composer has run. This seems fine to me and less work than committing the file unless you see an issue with that.

@roland-d @nikosdion Does that mean the files is fetched from Fido with every run of composer where it has not rune before or the file has been removed or it is older than 10 days (864000 seconds)? In this case: What do we do if Fido is not reachable when we build a release? In my opinion the file has to be in the repository. Otherwise we still have the problem of creating a DDoS, it's just a bit smaller.

Another thing: Maybe Fido currently is not down but blocked certain user agents or IP ranges due to our DDoS?

richard67 commented:
P.S.: We have composer install running on quite a bunch of docker images with every single commit of every single PR. If we do not have the file in our sources, we will fetch if every time because the docker images are new for each test. This still might cause them (or their server protection) to block these requests.

nikosdion commented:
@richard67 We talked about that with Harald yesterday, he told me he'd have drone.yaml to create a dummy, empty `plugins/system/webauthn/fido.jwt` file. This will indeed prevent the script from trying to get a new file which means the PR builds will work fine without hitting the FIDO server.

richard67 commented:
P.P.S.: Ah, wait,. it's only fetched when packages are built ... so less critical ... but I still think it should be in the sources.

richard67 commented:
Well maybe I am wrong and there is no problem, and I just did not understand it on the first look.

Regarding testing: I can test the performance thing, but I can not test if webauthn with Fido is still working (which in my opinion should be tested, too) because I have no Fido key.

nikosdion commented:
The WebAuthn part still works but it'd be great to have some people with Windows computers to test Windows Hello. Maybe @brianteeman if he's not super busy?

richard67 committed:
Update deleted files to PR #38664

richard67 commented:
@nikosdion As we do not have any other deleted files and folders from 4.2.1 to 4.2.2, I've decided to make a PR for you to make that change in script.php here in your PR: https://github.com/nikosdion/joomla-cms/pull/11

This will save the release lead one PR to be merged.

If you don't want that, let me know, and I will make a separate PR for that in this repo here.

Commit:
Merge pull request #11 from richard67/4.2-dev-nikosdion-fix-38233-webauth-preload-fido-deleted-file

[CMS PR 38664]

nikosdion commented:
Thank you, Richard! I just merged it :)

richard67 commented:
Thanks Nik.

richard67 committed:
Merge branch '4.2-dev' into fix/38233-webauth-preload-fido

richard67 commented:
I have tested this item :white_check_mark: successfully on a66514a63124009fd8ccc783c6cb82491a480ac0<br /><br />I have updated a 4.2.1 site (copy of my 3.10 homepage once updated to J4 for development) on my domain to the custom update URL provided by drone for this PR.

I could successfully verify that the performance problem is fixed by this PR.

In addition I have successfully tested that Webauthn still works as primary login authenticator and as MFA factor and both together. I have used Windows Hello with a PIN on my computer and Firefox latest stable.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38664">issues.joomla.org/tracker/joomla-cms/38664</a>.</sub>

tecpromotion commented:
I have tested this item :white_check_mark: successfully on a66514a63124009fd8ccc783c6cb82491a480ac0<br /><br />I was able to test it successfully on multiple websites and different hosts.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38664">issues.joomla.org/tracker/joomla-cms/38664</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38664">issues.joomla.org/tracker/joomla-cms/38664</a>.</sub>

roland-d committed:
Merge branch '4.2-dev' into fix/38233-webauth-preload-fido

roland-d commented:
Thank you
